### ISSUE TITLE
Problems while upgrading to Sourcegraph 3.0.1

### ISSUE THREAD

**Issue by ruhullahshah** [sourcegraph/sourcegraph#2259]
- **Sourcegraph version:** 3.0.1
- **Platform information:** Linux 

Hi, 

Thanks and congrats for getting in C# code intelligence support.

We were running Sourcegraph-2.8.9 and wanted to upgrade it to 3.0.1. 

After upgrading, we got the following message upon accessing the Sourcegraph server:
`Access to Sourcegraph is forbidden because no authentication provider is set in site configuration.`

We tried to downgrade to 2.8,9 to rectify the configuration, but seems like the changes made by 3.0.1 to the Sourcegraph configuration data are not compatible with 2.8.9. PFA the error log upon launching a 2.8.9 after upgrading to 3.0.1

Is there a workaround for this?

Thanks,
Ruh
[error_log_for_downgrade.txt](https://github.com/sourcegraph/sourcegraph/files/2854964/error_log_for_downgrade.txt)

**Comment by ruhullahshah** [comment]
Additional info:
`auth.provider` is set to `"builtin"` in `sourcegraph-config.json`

**Comment by slimsag** [comment]
Sounds like your auth provider from the old `sourcegraph-config.json` didn't get copied into the management console for some reason. Can you visit your management console (see https://docs.sourcegraph.com/admin/management_console) and check what that config looks like / that there is an `auth.provider` configured there?

See also https://docs.sourcegraph.com/admin/migration/3_0#configuration-is-now-stored-in-the-database-not-in-a-config-json-file

**Comment by sqs** [comment]
@ruhullahshah Could you please provide the info @slimsag 	mentioned above to help us solve this issue for you? Thanks!

**Comment by ruhullahshah** [comment]
@slimsag Thanks for the feedback. 
@slimsag  @sqs: I could not try out the suggestion yet, due to some infrastructural problems we ran into lately. I will keep this thread updated as soon as I get a chance to move further.

**Comment by ruhullahshah** [comment]
Hi. After fixing the configuration, as suggested by @slimsag, I am able to access the Sourcegraph instance through the browser. However, none of the users are able to login using their legitimate, existing credentials

The configuration looks like
`
{
  "auth.public": false,
  "auth.providers": [
    {
      "allowSignup": true,
      "type": "builtin"
    }
  ],
  "update.channel": "release"
}`

**Comment by slimsag** [comment]
That is very strange, apologies that you are running into this issue.

The only thing I can think of that may cause this is if somehow your instance is no longer connected to the same PostgreSQL database OR has somehow lost data.

To rule out this possibility, can you [grab a `psql` database prompt](https://docs.sourcegraph.com/admin/faq#how-do-i-access-the-sourcegraph-database) and confirm the user accounts exist in the database? e.g. by running:

```sql
SELECT username, created_at, passwd FROM users;
```

You should get some output like this:

```
stephen=# SELECT username, created_at, passwd FROM users;
 username |          created_at           |                            passwd                            
----------+-------------------------------+--------------------------------------------------------------
 slimsag  | 2019-01-17 07:43:12.951408+00 | $2a$10$asdwads123asdCoijAOijOijWwijoasdijWjoaisjd
(1 row)
```

**Do not upload your actual output here**, just confirm that the expected `username` shows up in the output and that the `passwd` field has something in it.

**Comment by ruhullahshah** [comment]
@slimsag I confirm that the users do exist. I  do not think that the user data is lost. On a side note, I was trying to downgrade to a previous version 2.x, to restore the server for now, but seems like it cannot read the volumes any more. Does Sourcegraph 3 change the existing configuration in so e way and makes it incompatible with the earlier versions? Does it create a backup of the existing configuration somewhere?

**Comment by slimsag** [comment]
Thanks for confirming!

The data should be completely backwards compatible, downgrading should work and you shouldn't be facing the error you are seeing when downgrading. I suspect the issue you are facing in 2.x is the same issue you are facing in 3.x. It seems like your postgres data directory may not be getting mounted into the container properly.

1. What does your `docker run` command look like (when starting either version)? Is it possible it is missing a volume mount like `--volume ~/.sourcegraph/data:/var/opt/sourcegraph`?
2. On the host machine, does your data directory contain a valid postgres data directory? e.g.:

```
cat ~/.sourcegraph/data/postgresql/PG_VERSION 
11
```

3. In the container, what does `/var/opt/sourcegraph/postgresql` look like? e.g. find the name of the docker container via `docker ps`, then run:

```
docker exec -it nostalgic_saha ls /var/opt/sourcegraph/postgresql
```

**Comment by ruhullahshah** [comment]
1. Yes the necessary volumes are mounted. We use `docker-compose up` to start Sourcegraph and the volumes we supply are:
`/export/sourcegraph/conf:/etc/sourcegraph`
`/export/sourcegraph/data:/var/opt/sourcegraph`

2. On host machine, running `cat ~/.sourcegraph/data/postgresql/PG_VERSION` returns 9.6

3. Running `ls /var/opt/sourcegraph/postgresql` on a Sourcegraph-3.0.1 instance returns a list of directories mostly pre-fixed with `pg_` and some `.conf` files.

I wanted to bring it to your notice, that while entering valid user credentials, I am returned to the login prompt in the user, but upon entering invalid credentials, I get an error message in the UI. This implies that my installation of Sourcegraph does differentiate between invalid and valid credentials, however, it does not proceed further. Also I do not see any console output in the Sourcegraph host machine upon successful authentication, whereas an invalid login attempt is reported to the console.

Regarding the downgrade, upon trying to launch 2.8.9, I get the following error log:
Attaching to sourcegraph
sourcegraph    | 10:01:04       postgres | FATAL:  "/var/opt/sourcegraph/postgresql" is not a valid data directory
sourcegraph    | 10:01:04       postgres | DETAIL:  File "/var/opt/sourcegraph/postgresql/PG_VERSION" does not contain valid data.
sourcegraph    | 10:01:04       postgres | HINT:  You might need to initdb.
sourcegraph    | 10:01:04       postgres | Terminating postgres
sourcegraph    | postgres died. Shutting down...
sourcegraph    | 10:01:04 syntect_server | Terminating syntect_server
sourcegraph    | 10:01:04      lsp-proxy | Terminating lsp-proxy
sourcegraph    | 10:01:04   repo-updater | Terminating repo-updater
sourcegraph    | 10:01:04        indexer | Terminating indexer
sourcegraph    | 10:01:04       frontend | Terminating frontend
sourcegraph    | 10:01:04      gitserver | Terminating gitserver
sourcegraph    | 10:01:04          redis | Terminating redis
sourcegraph    | 10:01:04   github-proxy | Terminating github-proxy
sourcegraph    | 10:01:04   query-runner | Terminating query-runner
sourcegraph    | 10:01:04        symbols | Terminating symbols
sourcegraph    | 10:01:04       searcher | Terminating searcher
sourcegraph exited with code 0

We did not use to observe this error before the attempted upgrade.

**Comment by slimsag** [comment]
@tsenart  have you ever seen anything like this downgrade error?

**Comment by slimsag** [comment]
@ruhullahshah Apologies for the trouble you are facing here, I've been trying to reproduce your issue to find you a quick solution, but unfortunately haven't been able to reproduce the login issue.

1. **_I was wrong about being able to downgrade from 3.x_**, if you are using the built-in PostgreSQL database and not a custom external one, downgrading to 2.8.9 again is not possible ([details here](https://docs.sourcegraph.com/admin/migration/3_0)). Apologies for my confusion here.
2. I believe that figuring out the user sign in problem you are facing in 3.x will be the quickest way to bring up Sourcegraph again.

Thoughts:

- Please run the latest sourcegraph/server v3.1.1, i.e. `sourcegraph/server:3.1.1`. Does the issue still occur?
- Can you sign in to the admin account, or are you also locked out of that?
- If you create a new plain user account, can you sign out and back in to that user account?
- What do the server logs look like? (you can email them to stephen@sourcegraph.com)

**Comment by ruhullahshah** [comment]
@slimsag No worries. Thanks for your support and suggestions
- The issue is observed in 3.1.1 as well.
- No I can't sign in to the admin account as well.
- Creating a new user account, seems to be working but again I am not able to log in.
- The console output of Sourcegraph is empty most of the time, except when I enter a wrong password

**Comment by ruhullahshah** [comment]
I have yet another observation. After trying to provide a clean slate (temporarily moving the `conf/` and `data/` to another location), and launching Sourcegraph, I am taken to a welcome screen for creating the admin account. Thereafter, when I try to log in with the newly created admin account, the same problem persists. i.e. being challenged again with a login screen. I tried this on v2.8.9, v3.0.1, and v3.1.1. Also I used the plain command available on the Sourcegraph website [Installation](https://docs.sourcegraph.com/admin/install/docker)

**Comment by slimsag** [comment]
Hi @ruhullahshah -- so sorry for the delay here. You being stuck without access to Sourcegraph for 15d is completely unacceptable.

The fact that you are running into this issue with even a clean state is VERY strange and baffling. Can you try running the `docker run` command without the `--volume` mounts? e.g.:

```bash
docker run --publish 7080:7080 --publish 2633:2633 --rm sourcegraph/server:3.1.1
```

After this, you should be able to create the admin account, sign out and back in just fine.

1. **If you can sign back in** -> This would lead me to believe something is wrong with the permissions on the volume mounts. I would need to confirm, but I believe if you are using Linux they may need to be chown'd to root.
2. **If you cannot sign back in** -> This would lead me to believe that something may be wrong with your browser storing the cookie OR for some reason the Redis instance inside of Sourcegraph is failing.

**Comment by ggilmore** [comment]
Hello @ruhullahshah,

Could you tell us:

How do you expose your instance to the public internet? (reverse proxy? directly expose the container port?) Did this change during your failed upgrade from `2.8.9` to `3.1.1`?

What is your `appURL`/`externalURL` setting in `sourcegraph-config.json` and the management console?

What browser are you using? What happens if you sign into our instance via Incognito / Private browsing mode?

Are there any errors that you see in the javascript console?

**Comment by ruhullahshah** [comment]
@slimsag Hi. No worries :)

I ran the command without the `--volume` mounts and could not log in back with a newly created account.  I guess you are right, something is up at browser side. See my answer to @ggilmore 

Hello @ggilmore 
1. No, we connect using HTTP only.
2. We directly expose the container port and it did not change from `2.8.9` to `3.1.1`.
3. We do not have this property set in `sourcegraph-config.json`
4. We are using Chrome `Version 72.0.3626.119` and I confirm that some of us are able to work with Sourcegraph via the Incognito mode.
5. There were no errors in the JS console.

In summary, Sourcegraph 3.1.1 works for us now with all the data we previously had, It is just that we never thought of switching to the Incognito mode and see if things work. Also I would like to report that for some of the users, running the same Chrome version as reported before, Sourcegraph works in a non-incognito tab as well. I am sure that there is a reason for this and would like to hear more about it from you guys.

@slimsag @ggilmore @sqs Thank you for your support and all the suggestions.

**Comment by slimsag** [comment]
Hi @ruhullahshah -- Apologies that we still haven't been able to reproduce this or find the cause on our end..

The fact that this works with an incognito window definitely indicates that _somehow_ the cookies currently in your browser are not compatible with 3.1 -- so clearing browser cookies should resolve the issue for a single user machine -- but we aren't aware of any changes that could have caused this. Me and a few others looked quite deeply into our handling of cookies and sessions and it all seems pretty robust to error states, so it is quite odd that you are hitting this honestly.

I know we've already asked for a lot of information, but could you answer:

1. In 3.x, `sourcegraph-config.json` is no longer used (at all), can you paste your exact configuration from the `/site-admin/configuration` page and management console (with any secrets redacted)?
2. I noticed you said "__some__ of us are able to work with Sourcegraph via incognito mode", have you observed any cases where incognito mode didn't work?
3. Can you confirm clearing browser cookies does resolve the issue on a machine which currently can't sign in outside of incognito? e.g.:

> - Paste `chrome://settings/siteData` into your URL bar
> - In the top right search bar (`Search cookies`), enter `sourcegraph.com`
> - Delete the entry labeled `sourcegraph.com`

I realize of course that if you have a lot of users that is not a good solution. We may unfortunately still need some more time to find the root cause here. It must be some cookie or session incompatibility between 2.x and 3.x.

**Comment by ruhullahshah** [comment]
Hi @slimsag Sorry for the delay in my reply to your query, but I was quite busy with some other tasks.
Thanks for all the investigation you guys did so far. 

1. Our `/site-admin/configuration`, looks like this:
```
{
  "disablePublicRepoRedirects": true,
  "email.address": "sourcegraph@<domain.name>",
  "email.smtp": {
    "authentication": "none",
    "domain": "<domain.name>",
    "host": "<domain.name>",
    "port": 25
  },
  "maxReposToSearch": 3000,
  "repoListUpdateInterval": 720
}
```

2.  No, I did not see a case wherein the incognito mode did not work. What I meant by saying that some of us can work with Sourcegraph in the incognito mode is that, there are some members who do not need to switch to incognito mode at all, and can work with a normal browser tab.

3. I confirm that deleting the Sourcegraph cookies resolves the issue and I no longer need to sign in from an incognito tab.

Thanks once again for all the suggestion and support.
